packages/client/src/internals/types.test.ts

```diff
@@ -0,0 +1,56 @@
+import { expectTypeOf } from 'expect-type';
+import isomorphicFetch from 'isomorphic-fetch';
+import nodeFetch from 'node-fetch';
+import type { fetch as undiciFetch } from 'undici';
+import { getFetch } from '../getFetch';
+import { getAbortController } from './fetchHelpers';
+import {
+  AbortControllerEsque,
+  AbortControllerInstanceEsque,
+  FetchEsque,
+  ResponseEsque,
+} from './types';
+
+describe('AbortController', () => {
+  test('AbortControllerEsque', () => {
+    expectTypeOf(
+      getAbortController,
+    ).returns.toEqualTypeOf<AbortControllerEsque | null>();
+
+    getAbortController(
+      null as unknown as typeof import('abort-controller')['AbortController'],
+    );
+
+    expectTypeOf(() => {
+      const AbortController = getAbortController(undefined)!;
+      return new AbortController();
+    }).returns.toEqualTypeOf<AbortControllerInstanceEsque>();
+  });
+});
+
+describe('fetch', () => {
+  test('FetchEsque', () => {
+    expectTypeOf(getFetch).returns.toEqualTypeOf<FetchEsque>();
+
+    expectTypeOf(() =>
+      getFetch()('', {
+        body: '',
+        headers: Math.random() > 0.5 ? [['a', 'b']] : { a: 'b' },
+        method: 'GET',
+        signal: new AbortSignal(),
+      }),
+    ).returns.toEqualTypeOf<Promise<ResponseEsque>>();
+
+    getFetch({} as unknown as typeof fetch);
+  });
+
+  test('NativeFetchEsque', () => {
+    getFetch(isomorphicFetch);
+    getFetch(nodeFetch);
+
+    // Passing in undiciFetch directly in Node v18.7.0 gives:
+    // ReferenceError: TextEncoder is not defined
+    // 🤷
+    getFetch({} as unknown as typeof undiciFetch);
```

Comment:
This one really stumped me 😕 

https://github.com/trpc/trpc/actions/runs/3873482321/jobs/6603548412

```plaintext
@trpc/tests:test-ci: FAIL client/src/internals/types.test.ts
@trpc/tests:test-ci:   ● Test suite failed to run
@trpc/tests:test-ci: 
@trpc/tests:test-ci:     ReferenceError: TextEncoder is not defined
@trpc/tests:test-ci: 
@trpc/tests:test-ci:       2 | import isomorphicFetch from 'isomorphic-fetch';
@trpc/tests:test-ci:       3 | import nodeFetch from 'node-fetch';
@trpc/tests:test-ci:     > 4 | import { fetch as undiciFetch } from 'undici';
@trpc/tests:test-ci:         | ^
@trpc/tests:test-ci:       5 | import { getFetch } from '../getFetch';
@trpc/tests:test-ci:       6 | import { getAbortController } from './fetchHelpers';
@trpc/tests:test-ci:       7 | import {
```

Node version is 18.7.0, which should support `TextEncoder`. I can repro the issue locally.